zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -93,35 +111,97 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
+    // [Root, NullifierS, CommitmentS, h_sig, h_iS, Residual Field Element(S)]
     // ie, below is the index mapping of the primary input elements on the protoboard:
     // - Index of the "Root" field elements: {0}
     // - Index of the "NullifierS" field elements: [1, NumInputs + 1[
     // - Index of the "CommitmentS" field elements: [NumInputs + 1, NumOutputs + NumInputs + 1[
-    // - Index of the "v_pub_in" field element: {NumOutputs + NumInputs + 1}
-    // - Index of the "v_pub_out" field element: {NumOutputs + NumInputs + 1 + 1}
+    // - Index of the "h_sig" field element: {NumOutputs + NumInputs + 1}
+    // - Index of the "Message Authentication TagS" (h_i) field elements: [NumOutputs + NumInputs + 1 + 1, NumOutputs + NumInputs + 1 + NumOuputs [
+    // - Index of the "Residual Field Element(s)" field elements: [NumOutputs + NumInputs + 1 + NumOuputs + 1 , NumOutputs + NumInputs + 1 + NumOuputs + 1 + nb_field_residual]
+    //
+    // The Residual field elements are structured as follows:
+    // - v_pub_in [0, size_value[
+    // - v_pub_out [size_value + 1, 2*size_value[
+    // - h_sig remaining bits [2*size_value + 1, 2*size_value + (digest_length-field_capacity)[
+    // - nullifierS remaining bits [2*size_value + (digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs)*(digest_length-field_capacity)[
+    // - commitmentS remaining bits [2*size_value + (1+NumInputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity)[
+    // - message authentication tagS remaining bits [2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+2*NumInputs+NumOutputs)*(digest_length-field_capacity)]
     // ============================================================================================ //
 
+    // This function extract the remaining bits of the nullifierS, commitmentS or h_iS from the residual field element(S)
+    function extract_extra_bits(uint start, uint length, uint[] memory primary_inputs) public view returns (bytes1) {
+        // We assume we extract less than 2 bytes
+        require(
+            length < 16,
+            "invalid range"
+        );
+
+        // If we do not want to extract any bits, return 0
+        if (length == 0) {
+            return bytes1(0x0);
+        }
+
+        // We first compute the offset (to forego the v_pubs and h_sig remaining bits)
+        uint offset = 2 * size_value + length_extra_bits;
+
+        // We then compute the position of the last bit to retrieve
+        uint end = start + length - 1;
+
+        // The remaining bits could be written over two field elements
+        // as such we compute the indices of the first and last field elements
+        // they are located at
+        uint index_start_bytes32 = 1 + 1 + 2*jsIn + jsOut + (offset+start)/field_capacity;
+        uint index_end_bytes32 = index_start_bytes32 + (length-1)/field_capacity;
```

Comment:
Since `length` is constrainted to be <16, `(length-1)/field_capacity` will always be 0, and so `index_end_bytes32 = index_start_bytes32`, right?

Reply:
You should do `uint index_end_bytes32 = 1 + 1 + 2*jsIn + jsOut + (offset+start+length-1)/field_capacity;` if I get your intention correctly. In fact, `(offset+start)/field_capacity` will be div_floored, and `(length-1)/field_capacity` = 0

Reply:
Yes you are right.

I tried to do an optimization and the tests passed only because of our particular case. I'll correct that right away.

Reply:
Let's have something working first, and we'll think about optimizations in a second time. For now, we should focus on having a **documented and clear code** that works

Reply:
https://github.com/clearmatics/zeth/pull/104